server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -480,4 +481,61 @@ public void onFailure(Exception e) {
             }
         };
     }
+
+    public void deleteDecommissionState(final ActionListener<ClusterStateUpdateResponse> listener) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+        DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
+
+        if (decommissionAttributeMetadata == null) {
+            throw new IllegalStateException("No decommission attribute found. Not a valid request");
+        }
+
+        String attributeName = decommissionAttributeMetadata.decommissionAttribute().attributeName();
+        // Set the weights back to 1 for all the zones
+        List<String> awarenessValues = forcedAwarenessAttributes.get(attributeName);
+        Map<String, Double> weights = new HashMap<>();
+        awarenessValues.forEach(awarenessValue -> { weights.put(awarenessValue, 1.0); });
+        decommissionController.setWeights(attributeName, weights, new ActionListener<>() {
+            @Override
+            public void onResponse(ClusterPutWeightedRoutingResponse clusterPutWeightedRoutingResponse) {
+                // Weights are successfully set. Let clear the decommission attribute.
+                clearDecommissionAttribute(listener);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error(() -> new ParameterizedMessage("Failed to set the weights for removing decommission."), e);
+                listener.onResponse(new ClusterStateUpdateResponse(false));
+            }
+        });
+    }
+
+    private void clearDecommissionAttribute(final ActionListener<ClusterStateUpdateResponse> listener) {
+        clusterService.submitStateUpdateTask("delete_decommission_state", new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                return clearDecommissionClusterState(currentState);
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                logger.error(() -> new ParameterizedMessage("Failed to clear decommission attribute."), e);
+                listener.onResponse(new ClusterStateUpdateResponse(false));
```

Comment:
If the setting of weight fails then the decommissionStatus will not be cleared. And we are responding with acknowledged status as false to signify that the request to clear the decommission has failed.

Reply:
I understand that we are acknowledging with `false` but then we are expecting user to retry.

Generally, the expectation from APIs that mutate the resource is: it will move the state from `A` to `B` and if it fails, the original state `A` remains. With current implementation, it is possible that we will not go back to `A` but the state will be `A'`.

Want to understand if it is expected from this API.

Reply:
Yes. We are currently expecting the operator to do a retry. If they are trying for zone recommission.

Reply:
I would like to know reasons behind not making it atomic. I understand that it may not be possible to make it completely atomic (what if the reverting back weights fails) but we should at least have best efforts solution.

I am fine if you want to discuss this more by creating a tracking issue.